src/index.js

```diff
@@ -1,38 +1,76 @@
     /**
      * handle_kernel 
-     * creates an instance of a "Manager" used to listen for new comms and create new components
+     * registers comm targets with the kernel comm_manager
+     * when new comms are open, renders a Parent component that takes over rendering of actual components
      */
     const handle_kernel = ( Jupyter, kernel ) => {
-      if ( kernel.comm_manager && kernel.component_manager === undefined ) {
-        kernel.component_manager = new Manager.ComponentManager( kernel, Comm );
-      } 
+      // register the target comm / listens for new comms 
+      kernel.comm_manager.register_target( commTarget, ( comm, msg ) => {
+        if ( msg[ 'msg_type' ] === 'comm_open' ) {
+          const msg_id = msg.parent_header.msg_id;
+          const cell = Jupyter.notebook.get_msg_cell( msg_id );
+            
+          if ( cell.react_output && cell.react_output[ commTarget ] ) {
+            const component = React.createElement( Component,  { ...componentParams, comm, comm_msg: msg } );
+            ReactDom.render( component, cell.react_output[ commTarget ].subarea );
+          }
+        }
+      });
 
-      if ( kernel.component_manager ) {
-        const Component = ReactComponent( componentParams );
-        kernel.component_manager.register( commTarget, Component ) 
-      }
+      // find any open comms and render components 
+      kernel.comm_info( commTarget, commInfo => {
+        const comms = Object.keys( commInfo[ 'content' ][ 'comms' ] );
+        const md = Jupyter.notebook.metadata;
+
+        if ( comms.length && md.react_comms ) {
+          comms
+            .filter( c => md.react_comms[ c.comm_id ] && c )
+            .forEach( id => {
+              const cell = Jupyter.notebook.get_cells()[ parseInt( md.react_comms[ id ] ) ];
+              if ( cell ) {
+                const module = id.split( '.' ).slice( -1 )[ 0 ];
+                const newComm = new Comm.Comm( commTarget, id );
+                kernel.comm_manager.register_comm( newComm );
+
+                const component = React.createElement( Component,  { ...componentParams, comm: newComm, comm_msg: { content: { data: { module } } } } );
+                ReactDom.render( component, cell.react_output[ commTarget ].subarea );
+              }
+            });
+        }
+    })
     };
 
     /**
      * handle_cell 
      * add react dom area for components to render themselves into 
      * @param {object} notebook cell
      */
-    // TODO need to handle clear out output calls
     const handle_cell = ( cell ) => {
       if ( cell.cell_type === 'code' ) {
-        if ( !cell.react_dom ) {
-          cell.react_dom = new Area( cell );
-        } else if ( cell.react_dom.clear !== undefined ) {
-          cell.react_dom.clear();
+        if ( !cell.react_output ) {
+          cell.react_output = {};
         }
+
+        if ( !cell.react_output[ commTarget ] ) {
+          cell.react_output[ commTarget ] = new Output( cell );
```

Comment:
@chelm does this mean you create divs for every cell whether or not the component will be rendered there? if so, does it need to be that way instead of creating and destroying the divs on the fly?